core/dbt/parser/base.py

```diff
@@ -306,6 +306,19 @@ def update_parsed_node_config(
             else:
                 parsed_node.docs = Docs(show=docs_show)
 
+        # If we have constraints_enabled in the config, copy to node level, for backwards
+        # compatibility with earlier node-only config.
+        if config_dict.get("constraints_enabled", False):
+            parsed_node.constraints_enabled = True
+
+            parser_name = type(self).__name__
+            if parser_name == "ModelParser":
+                original_file_path = parsed_node.original_file_path
+                error_message = "\n    `constraints_enabled=true` can only be configured within `schema.yml` files\n      NOT within a model file(ex: .sql, .py) or `dbt_project.yml`."
+                raise ParsingError(
+                    f"Original File Path: ({original_file_path})\nConstraints must be defined in a `yml` schema configuration file like `schema.yml`.\nOnly the SQL table materialization is supported for constraints. \n`data_type` values must be defined for all columns and NOT be null or blank.{error_message}"
+                )
```

Comment:
Confirming that this is necessary only:
- in order to enable parse-time validation that every column for this model has a `data_type` defined
- because of this tech debt ticket: https://github.com/dbt-labs/dbt-core/issues/6511

As @gshank commented earlier, parse time is the right time for this sort of validation, versus compile-time, per our slightly nonstandard meanings of "parsing" + "compilation." I do see the downside of, we return parse-time errors as soon as we hit them, rather than batching up all the errors we can and returning them all. If it were a _warning_ instead, then we could return them all together!

FWIW - I'm aligned with `constraints_enabled` being a _config_, with the idea that it should be possible to enable a broad swath of models (e.g. "everything in the `marts/public` folder") to have type-checked column specs. That's true even though it depends on the `columns` attribute, which is _not_ a config, but a property specific to one model.

If we can't resolve that tech debt before the v1.5 release, I'd probably opt for removing the parse-time check, and letting the database return its own validation error at runtime. It's a less-nice UX to see the error later in the dev cycle, but I think it's better to offer the full flexibility around configuring `constraints_enabled`.

Reply:
You are correct on why this is a Parsing Error today. I'm fine with the less-nice UX in favor of `constraints_enabled=true` being available in all configs come v1.5.